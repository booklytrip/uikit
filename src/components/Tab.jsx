/**
 * @flow
 */

import React, { Component } from 'react';
import classNames from 'classnames';
import { map } from 'lodash';

import { Menu, Loader } from 'semantic-ui-react';

type TabMenuItemProps = {
    className: string,
    onClick: Function,
    label: string,
    index: number,
    active?: boolean,
};

type TabProps = {
    className: string,
    label: string,
    style?: Object,
    children: any,
    loading?: boolean,
    active?: boolean,
};

type TabsProps = {
    className: string,
    children: any,
    menuItemComponent: Object,
    tabular?: boolean,
    secondary?: boolean,
    pointing?: boolean,
};

const styles = {
    loader: {
        marginTop: '20px',
        marginBottom: '20px',
    },
};

/**
 * Single menu item element
 *
 * @param {String}   label   - An item label
 * @param {Number}   value   - A value of item
 * @param {Bool}     active  - Activates current menu item if set as true
 * @param {Function} onClick - A callback to trigger when item is clicked
 */
const TabMenuItem = ({
    className,
    label,
    value,
    active,
    onClick,
}: TabMenuItemProps) => (
    <Menu.Item
        onClick={onClick}
        className={className}
        active={active}
        value={value}
        link
    >
        {label}
    </Menu.Item>
);

/**
 * A single tab element
 *
 * @param {Bool}       loading  - Show loader for this tab
 * @param {Bool}       active   - Activates current tab if set as true
 * @param {Array|Node} children - A content to present inside tab
 * @param {Object}     style    - An object with styles applied to the tab
 */
const Tab = ({ className, loading, active, children, style }: TabProps) => (
    <div
        className={classNames('ui bottom attached tab segment', className, {
            active,
        })}
        style={style}
    >
        {active && loading
            ? <Loader size="big" style={styles.loader} inline centered />
            : children}
    </div>
);

/**
 * Tabs container manage internal state of active tab
 */
class Tabs extends Component {
    props: TabsProps;

    static defaultProps: {
        menuItemComponent: Object,
    };

    state: {
        active: number,
    };

    constructor(props: TabsProps) {
        super(props);

        this.state = {
            active: 0,
        };
    }

    onClick = (e, { value }) => {
        this.setState({ active: value });
    };

    render() {
        const { active } = this.state;

        const {
            className,
            children,
            menuItemComponent,
            secondary,
            pointing,
        } = this.props;

        const labels = React.Children.map(children, (child, idx) => ({
            key: idx,
            value: idx,
            label: child.props.label,
            active: active === idx,
        }));

        const MenuItemComponent = menuItemComponent;

        return (
            <div>
                <Menu
                    className={className}
                    items={labels}
                    activeIndex={active}
                    attached="top"
                    secondary={secondary}
                    pointing={pointing}
                    tabular
                >
                    {map(labels, label => (
                        <MenuItemComponent
                            key={label.key}
                            {...label}
                            onClick={this.onClick}
                        />
                    ))}
                </Menu>
                {React.Children.map(children, (child, idx) =>
                    React.cloneElement(child, {
                        active: active === idx,
                    }),
                )}
            </div>
        );
    }
}

Tabs.defaultProps = {
    menuItemComponent: TabMenuItem,
};

export { Tabs, Tab };
