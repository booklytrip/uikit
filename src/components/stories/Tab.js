import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@kadira/react-storybook-decorator-centered';

import { Tabs, Tab } from '../Tab.jsx';

storiesOf('Tab', module)
    .addDecorator(centered)
    .add('one tab', () => (
        <Tabs>
            <Tab label="First">
                First tab
            </Tab>
        </Tabs>
    ))
    .add('few tabs', () => (
        <Tabs>
            <Tab label="First">
                First tab
            </Tab>
            <Tab label="Second">
                Second tab
            </Tab>
            <Tab label="Third">
                Third tab
            </Tab>
        </Tabs>
    ));
