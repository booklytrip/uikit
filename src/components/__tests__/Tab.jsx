import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { Tabs, Tab } from '../Tab';

describe('Tab', () => {
    it('renders without crashing', () => {
        mount(<Tab />);
    });
    
    describe('Snapshots', () => {
        it('renders in loading state', () => {
            const rendered = renderer.create(<Tab loading>Tab</Tab>);
            expect(rendered.toJSON()).toMatchSnapshot();
        });
        
        it('renders in active state', () => {
            const rendered = renderer.create(<Tab active>Tab</Tab>);
            expect(rendered.toJSON()).toMatchSnapshot();
        });
    });
});

describe('Tabs', () => {
    it('renders without crashing', () => {
        mount(<Tabs />);
    });
    
    describe('Snapshots', () => {
        it('renders with multiple tabs', () => {
            const rendered = renderer.create(
                <Tabs>
                    <Tab>First</Tab>
                    <Tab>Second</Tab>
                </Tabs>
            );
            expect(rendered.toJSON()).toMatchSnapshot();
        });
        
        it('renders formated menu with tabs', () => {
            const rendered = renderer.create(
                <Tabs tabular>
                    <Tab>First</Tab>
                    <Tab>Second</Tab>
                </Tabs>
            );
            expect(rendered.toJSON()).toMatchSnapshot();
        });
        
        it('renders menu as a secondary', () => {
            const rendered = renderer.create(
                <Tabs secondary>
                    <Tab>First</Tab>
                    <Tab>Second</Tab>
                </Tabs>
            );
            expect(rendered.toJSON()).toMatchSnapshot();
        });

        it('renders with a pointing menu', () => {
            const rendered = renderer.create(
                <Tabs pointing>
                    <Tab>First</Tab>
                    <Tab>Second</Tab>
                </Tabs>
            );
            expect(rendered.toJSON()).toMatchSnapshot();
        });
    });
});
