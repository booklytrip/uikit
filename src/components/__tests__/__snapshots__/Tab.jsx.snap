exports[`Tab Snapshots renders in active state 1`] = `
<div
  className="ui bottom attached tab segment active"
  style={undefined}>
  Tab
</div>
`;

exports[`Tab Snapshots renders in loading state 1`] = `
<div
  className="ui bottom attached tab segment"
  style={undefined}>
  Tab
</div>
`;

exports[`Tabs Snapshots renders formated menu with tabs 1`] = `
<div>
  <div
    className="ui menu tabular top attached">
    <a
      className="item active"
      onClick={[Function]} />
    <a
      className="item"
      onClick={[Function]} />
  </div>
  <div
    className="ui bottom attached tab segment active"
    style={undefined}>
    First
  </div>
  <div
    className="ui bottom attached tab segment"
    style={undefined}>
    Second
  </div>
</div>
`;

exports[`Tabs Snapshots renders menu as a secondary 1`] = `
<div>
  <div
    className="ui menu secondary tabular top attached">
    <a
      className="item active"
      onClick={[Function]} />
    <a
      className="item"
      onClick={[Function]} />
  </div>
  <div
    className="ui bottom attached tab segment active"
    style={undefined}>
    First
  </div>
  <div
    className="ui bottom attached tab segment"
    style={undefined}>
    Second
  </div>
</div>
`;

exports[`Tabs Snapshots renders with a pointing menu 1`] = `
<div>
  <div
    className="ui menu tabular pointing top attached">
    <a
      className="item active"
      onClick={[Function]} />
    <a
      className="item"
      onClick={[Function]} />
  </div>
  <div
    className="ui bottom attached tab segment active"
    style={undefined}>
    First
  </div>
  <div
    className="ui bottom attached tab segment"
    style={undefined}>
    Second
  </div>
</div>
`;

exports[`Tabs Snapshots renders with multiple tabs 1`] = `
<div>
  <div
    className="ui menu tabular top attached">
    <a
      className="item active"
      onClick={[Function]} />
    <a
      className="item"
      onClick={[Function]} />
  </div>
  <div
    className="ui bottom attached tab segment active"
    style={undefined}>
    First
  </div>
  <div
    className="ui bottom attached tab segment"
    style={undefined}>
    Second
  </div>
</div>
`;
